// relay-compiler/lib/RelayCompilerPublic.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */

'use strict';

var _require = require('graphql-compiler'),
    GraphQLCompilerContext = _require.CompilerContext;

var _require2 = require('graphql-compiler'),
    ASTConvert = _require2.ASTConvert,
    CodegenRunner = _require2.CodegenRunner,
    ConsoleReporter = _require2.ConsoleReporter,
    MultiReporter = _require2.MultiReporter;

module.exports = {
  ConsoleReporter: ConsoleReporter,
  Parser: require('./RelayParser'),
  CodeGenerator: require('./RelayCodeGenerator'),

  GraphQLCompilerContext: GraphQLCompilerContext,

  /** @deprecated Use JSModuleParser. */
  FileIRParser: require('./RelayJSModuleParser'),

  FileWriter: require('./RelayFileWriter'),
  IRTransforms: require('./RelayIRTransforms'),
  JSModuleParser: require('./RelayJSModuleParser'),
  MultiReporter: MultiReporter,
  Runner: CodegenRunner,
  compileRelayArtifacts: require('./compileRelayArtifacts'),
  formatGeneratedModule: require('./formatGeneratedModule'),
  convertASTDocuments: ASTConvert.convertASTDocuments,
  transformASTSchema: ASTConvert.transformASTSchema
};
