// relay-compiler/lib/writeRelayGeneratedFile.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */

'use strict';

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _extends3 = _interopRequireDefault(require('babel-runtime/helpers/extends'));

let writeRelayGeneratedFile = (() => {
  var _ref = (0, _asyncToGenerator3.default)(function* (codegenDir, generatedNode, formatModule, flowText, _persistQuery, platform, relayRuntimeModule, sourceHash) {
    // Copy to const so Flow can refine.
    var persistQuery = _persistQuery;
    var moduleName = generatedNode.name + '.graphql';
    var platformName = platform ? moduleName + '.' + platform : moduleName;
    var filename = platformName + '.js';
    var flowTypeName = generatedNode.kind === RelayConcreteNode.FRAGMENT ? 'ConcreteFragment' : generatedNode.kind === RelayConcreteNode.REQUEST ? 'ConcreteRequest' : generatedNode.kind === RelayConcreteNode.BATCH_REQUEST ? 'ConcreteBatchRequest' : 'empty';
    var devOnlyProperties = {};

    var docText = void 0;
    if (generatedNode.kind === RelayConcreteNode.REQUEST) {
      docText = generatedNode.text;
    } else if (generatedNode.kind === RelayConcreteNode.BATCH_REQUEST) {
      docText = generatedNode.requests.map(function (request) {
        return request.text;
      }).join('\n\n');
    }

    var hash = null;
    if (generatedNode.kind === RelayConcreteNode.REQUEST || generatedNode.kind === RelayConcreteNode.BATCH_REQUEST) {
      var oldHash = Profiler.run('RelayFileWriter:compareHash', function () {
        var oldContent = codegenDir.read(filename);
        // Hash the concrete node including the query text.
        var hasher = require('crypto').createHash('md5');
        hasher.update('cache-breaker-7');
        hasher.update(JSON.stringify(generatedNode));
        hasher.update(sourceHash);
        if (flowText) {
          hasher.update(flowText);
        }
        if (persistQuery) {
          hasher.update('persisted');
        }
        hash = hasher.digest('hex');
        return extractHash(oldContent);
      });
      if (hash === oldHash) {
        codegenDir.markUnchanged(filename);
        return null;
      }
      if (codegenDir.onlyValidate) {
        codegenDir.markUpdated(filename);
        return null;
      }
      if (persistQuery) {
        switch (generatedNode.kind) {
          case RelayConcreteNode.REQUEST:
            devOnlyProperties.text = generatedNode.text;
            generatedNode = (0, _extends3['default'])({}, generatedNode, {
              text: null,
              id: yield persistQuery(require('fbjs/lib/nullthrows')(generatedNode.text))
            });
            break;
          case RelayConcreteNode.BATCH_REQUEST:
            devOnlyProperties.requests = generatedNode.requests.map(function (request) {
              return {
                text: request.text
              };
            });
            generatedNode = (0, _extends3['default'])({}, generatedNode, {
              requests: yield Promise.all(generatedNode.requests.map((() => {
                var _ref2 = (0, _asyncToGenerator3.default)(function* (request) {
                  return (0, _extends3['default'])({}, request, {
                    text: null,
                    id: yield persistQuery(require('fbjs/lib/nullthrows')(request.text))
                  });
                });

                return function (_x9) {
                  return _ref2.apply(this, arguments);
                };
              })()))
            });
            break;
          case RelayConcreteNode.FRAGMENT:
            // Do not persist fragments.
            break;
          default:
            generatedNode.kind;
        }
      }
    }

    var devOnlyAssignments = require('./deepMergeAssignments')('node', devOnlyProperties);

    var moduleText = formatModule({
      moduleName: moduleName,
      documentType: flowTypeName,
      docText: docText,
      flowText: flowText,
      hash: hash ? '@relayHash ' + hash : null,
      concreteText: require('./dedupeJSONStringify')(generatedNode),
      devOnlyAssignments: devOnlyAssignments,
      relayRuntimeModule: relayRuntimeModule,
      sourceHash: sourceHash
    });

    codegenDir.writeFile(filename, moduleText);
    return generatedNode;
  });

  return function writeRelayGeneratedFile(_x, _x2, _x3, _x4, _x5, _x6, _x7, _x8) {
    return _ref.apply(this, arguments);
  };
})();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _require = require('relay-runtime'),
    RelayConcreteNode = _require.RelayConcreteNode;

var _require2 = require('graphql-compiler'),
    Profiler = _require2.Profiler;

/**
 * Generate a module for the given document name/text.
 */


function extractHash(text) {
  if (!text) {
    return null;
  }
  if (/<<<<<|>>>>>/.test(text)) {
    // looks like a merge conflict
    return null;
  }
  var match = text.match(/@relayHash (\w{32})\b/m);
  return match && match[1];
}

module.exports = writeRelayGeneratedFile;
