// chrome-debugging-protocol/lib/domain/page.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var chrome_debugging_domain_1 = require("../chrome-debugging-domain");
var PageDomain = (function (_super) {
    __extends(PageDomain, _super);
    function PageDomain() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // methods
    PageDomain.prototype.enable = function () {
        return this.send('enable');
    };
    PageDomain.prototype.disable = function () {
        return this.send('disable');
    };
    PageDomain.prototype.reload = function (params) {
        return this.send('reload', params);
    };
    PageDomain.prototype.navigate = function (params) {
        return this.send('navigate', params);
    };
    PageDomain.prototype.handleJavaScriptDialog = function (params) {
        return this.send('handleJavaScriptDialog', params);
    };
    PageDomain.prototype.configureOverlay = function (params) {
        return this.send('configureOverlay', params);
    };
    // events
    PageDomain.prototype.domContentEventFired = function (cb) {
        return this.addListener('domContentEventFired', cb);
    };
    PageDomain.prototype.loadEventFired = function (cb) {
        return this.addListener('loadEventFired', cb);
    };
    PageDomain.prototype.frameAttached = function (cb) {
        return this.addListener('frameAttached', cb);
    };
    PageDomain.prototype.frameNavigated = function (cb) {
        return this.addListener('frameNavigated', cb);
    };
    PageDomain.prototype.frameDetached = function (cb) {
        return this.addListener('frameDetached', cb);
    };
    PageDomain.prototype.javascriptDialogOpening = function (cb) {
        return this.addListener('javascriptDialogOpening', cb);
    };
    PageDomain.prototype.javascriptDialogClosed = function (cb) {
        return this.addListener('javascriptDialogClosed', cb);
    };
    PageDomain.prototype.interstitialShown = function (cb) {
        return this.addListener('interstitialShown', cb);
    };
    PageDomain.prototype.interstitialHidden = function (cb) {
        return this.addListener('interstitialHidden', cb);
    };
    PageDomain.prototype.navigationRequested = function (cb) {
        return this.addListener('navigationRequested', cb);
    };
    return PageDomain;
}(chrome_debugging_domain_1.ChromeDebuggingDomain));
exports.PageDomain = PageDomain;
