// chrome-debugging-protocol/lib/domain/profiler.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var chrome_debugging_domain_1 = require("../chrome-debugging-domain");
var ProfilerDomain = (function (_super) {
    __extends(ProfilerDomain, _super);
    function ProfilerDomain() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // methods
    ProfilerDomain.prototype.enable = function () {
        return this.send('enable');
    };
    ProfilerDomain.prototype.disable = function () {
        return this.send('disable');
    };
    ProfilerDomain.prototype.setSamplingInterval = function (params) {
        return this.send('setSamplingInterval', params);
    };
    ProfilerDomain.prototype.start = function () {
        return this.send('start');
    };
    ProfilerDomain.prototype.stop = function () {
        return this.send('stop');
    };
    // events
    ProfilerDomain.prototype.consoleProfileStarted = function (cb) {
        return this.addListener('consoleProfileStarted', cb);
    };
    ProfilerDomain.prototype.consoleProfileFinished = function (cb) {
        return this.addListener('consoleProfileFinished', cb);
    };
    return ProfilerDomain;
}(chrome_debugging_domain_1.ChromeDebuggingDomain));
exports.ProfilerDomain = ProfilerDomain;
