// xatom-debug-chrome-base/lib/debugger.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var chrome_debugging_protocol_1 = require("chrome-debugging-protocol");
var path_1 = require("path");
var url_1 = require("url");
var fs_1 = require("fs");
var http_1 = require("http");
var https_1 = require("https");
var lodash_1 = require("lodash");
var SourceMapConsumer = require('source-map').SourceMapConsumer;
var ChromeDebuggingProtocolDebugger = (function () {
    function ChromeDebuggingProtocolDebugger() {
        this.connected = false;
        this.paused = false;
        this.breakpoints = [];
        this.scripts = [];
        this.callFrames = [];
        this.events = new events_1.EventEmitter();
        this.ignoreUrls = [
            '',
            'extensions::app'
        ];
    }
    // Methods
    ChromeDebuggingProtocolDebugger.prototype.disconnect = function () {
        if (this.protocol) {
            this.protocol.disconnect();
        }
        this.protocol = null;
        this.connected = false;
        this.breakpoints = [];
        this.scripts = [];
    };
    // override with debugger
    ChromeDebuggingProtocolDebugger.prototype.didConnect = function (domains) {
        return Promise.resolve();
    };
    ChromeDebuggingProtocolDebugger.prototype.getFilePathFromUrl = function (fileUrl) {
        return fileUrl;
    };
    ChromeDebuggingProtocolDebugger.prototype.fileExists = function (filePath) {
        return new Promise(function (resolve, reject) {
            fs_1.stat(filePath, function (err, stats) {
                if (err) {
                    reject(err);
                }
                else {
                    resolve(stats);
                }
            });
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.connect = function (socketUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var _a, _b, Console, Profiler, Runtime, Debugger, Page;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        this.protocol = new chrome_debugging_protocol_1.ChromeDebuggingProtocol(socketUrl);
                        _a = this;
                        return [4 /*yield*/, this.protocol
                                .connect()
                                .then(function (domains) {
                                _this.connected = true;
                                return domains;
                            })];
                    case 1:
                        _a.domains = _c.sent();
                        _b = this.domains, Console = _b.Console, Profiler = _b.Profiler, Runtime = _b.Runtime, Debugger = _b.Debugger, Page = _b.Page;
                        this.protocol.didClose(function () { return _this.events.emit('didClose'); });
                        // Add Listeners
                        Runtime.exceptionThrown(function (params) {
                            if (params.exceptionDetails) {
                                _this.events.emit('didThrownException', params);
                                var errorObject = {
                                    type: 'string',
                                    value: lodash_1.get(params, 'exceptionDetails.exception.description')
                                };
                                if (lodash_1.get(params, 'exceptionDetails.exception')) {
                                    errorObject = params.exceptionDetails.exception;
                                }
                                _this.events.emit('didLogMessage', {
                                    type: 'error',
                                    args: [errorObject]
                                });
                            }
                        });
                        // Page.loadEventFired((params) => {
                        //   this.scripts = []
                        // })
                        Runtime.consoleAPICalled(function (params) {
                            _this.events.emit('didLogMessage', params);
                        });
                        Debugger.paused(function (params) {
                            if (_this.skipFirstPause === true) {
                                Debugger.resume();
                                _this.skipFirstPause = false;
                            }
                            else {
                                _this.callFrames = params.callFrames;
                                _this.paused = true;
                                _this.events.emit('didPause', params);
                            }
                        });
                        Debugger.resumed(function (params) {
                            _this.paused = false;
                            _this.events.emit('didResume');
                        });
                        Debugger.scriptParsed(function (params) { return __awaiter(_this, void 0, void 0, function () {
                            var _this = this;
                            var isIgnored, script, mappingUrl, smc_1, rawSourcemap_1, sourcePath_1, isBase64, base64Content, mappingPath;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        isIgnored = this.ignoreUrls['includes'](String(params.url));
                                        if (isIgnored)
                                            return [2 /*return*/];
                                        params.originalUrl = params.url;
                                        params.url = this.getFilePathFromUrl(params.url);
                                        script = {
                                            scriptId: params.scriptId,
                                            url: params.url,
                                            sourceMapURL: params.sourceMapURL
                                        };
                                        if (!params.sourceMapURL) return [3 /*break*/, 6];
                                        sourcePath_1 = path_1.parse(params.url);
                                        isBase64 = params
                                            .sourceMapURL
                                            .match(/^data\:application\/json\;(charset=.+)?base64\,(.+)$/);
                                        if (!isBase64) return [3 /*break*/, 2];
                                        base64Content = window.atob(String(isBase64[2]));
                                        return [4 /*yield*/, this.getObjectFromString(base64Content)];
                                    case 1:
                                        rawSourcemap_1 = _a.sent();
                                        return [3 /*break*/, 5];
                                    case 2:
                                        mappingPath = path_1.join(sourcePath_1.dir, params.sourceMapURL);
                                        return [4 /*yield*/, this
                                                .getObjectFromFile(mappingPath)
                                                .catch(function () {
                                                mappingUrl = url_1.resolve(params.originalUrl, params.sourceMapURL);
                                            })];
                                    case 3:
                                        rawSourcemap_1 = _a.sent();
                                        if (!(mappingUrl && lodash_1.isUndefined(rawSourcemap_1))) return [3 /*break*/, 5];
                                        return [4 /*yield*/, this
                                                .getObjectFromUrl(mappingUrl)
                                                .catch(function (e) {
                                                // skip: Unable to get sourcemaps.
                                            })];
                                    case 4:
                                        rawSourcemap_1 = _a.sent();
                                        _a.label = 5;
                                    case 5:
                                        if (rawSourcemap_1) {
                                            if (lodash_1.get(rawSourcemap_1, 'sources')) {
                                                rawSourcemap_1.sources.forEach(function (sourceUrl, index) { return __awaiter(_this, void 0, void 0, function () {
                                                    var targetUrl;
                                                    return __generator(this, function (_a) {
                                                        targetUrl = this.getFilePathFromUrl(sourceUrl);
                                                        if (targetUrl === sourceUrl) {
                                                            targetUrl = path_1.join(sourcePath_1.dir, sourceUrl);
                                                        }
                                                        rawSourcemap_1.sources[index] = targetUrl;
                                                        return [2 /*return*/];
                                                    });
                                                }); });
                                            }
                                            smc_1 = new SourceMapConsumer(rawSourcemap_1);
                                            script.sourceMap = {
                                                getOriginalPosition: function (lineNumber, columnNumber) {
                                                    var lookup = {
                                                        line: lineNumber + 1,
                                                        column: columnNumber || 0,
                                                        bias: SourceMapConsumer.LEAST_UPPER_BOUND
                                                    };
                                                    var position = smc_1.originalPositionFor(lookup);
                                                    if (position.source === null) {
                                                        lookup.bias = SourceMapConsumer.GREATEST_LOWER_BOUND;
                                                        position = smc_1.originalPositionFor(lookup);
                                                    }
                                                    if (position.source) {
                                                        return {
                                                            url: position.source,
                                                            lineNumber: position.line - 1,
                                                            columnNumber: position.column
                                                        };
                                                    }
                                                    else {
                                                        return false;
                                                    }
                                                }
                                            };
                                            smc_1.sources.forEach(function (sourceUrl) {
                                                var mapScript = {
                                                    // scriptId: params.scriptId,
                                                    url: sourceUrl,
                                                    sourceMap: {
                                                        getPosition: function (lineNumber, columnNumber) {
                                                            var lookup = {
                                                                source: sourceUrl,
                                                                line: lineNumber + 1,
                                                                column: columnNumber || 0,
                                                                bias: SourceMapConsumer.LEAST_UPPER_BOUND
                                                            };
                                                            var position = smc_1.generatedPositionFor(lookup);
                                                            if (position.line === null) {
                                                                lookup.bias = SourceMapConsumer.GREATEST_LOWER_BOUND;
                                                                position = smc_1.generatedPositionFor(lookup);
                                                            }
                                                            return {
                                                                url: params.url,
                                                                lineNumber: position.line - 1
                                                            };
                                                        }
                                                    }
                                                };
                                                _this.addParsedScript(mapScript);
                                            });
                                        }
                                        else {
                                            atom.notifications.addError('XAtom Debug: Unable to map file', {
                                                detail: mappingUrl,
                                                dismissable: true
                                            });
                                        }
                                        _a.label = 6;
                                    case 6:
                                        this.addParsedScript(script);
                                        return [2 /*return*/];
                                }
                            });
                        }); });
                        return [4 /*yield*/, this.didConnect(this.domains)];
                    case 2: 
                    // trigger connected
                    return [2 /*return*/, _c.sent()];
                }
            });
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.addParsedScript = function (script) {
        var parsed = lodash_1.find(this.scripts, {
            url: script.url
        });
        if (parsed) {
            script = lodash_1.extend({}, parsed, script);
        }
        this.scripts.push(script);
        this.events.emit('didLoadScript', script);
    };
    ChromeDebuggingProtocolDebugger.prototype.getUrlForMappedSource = function (fileUrl) {
        return null;
    };
    ChromeDebuggingProtocolDebugger.prototype.getObjectFromUrl = function (fileUrl) {
        var _this = this;
        var urlParts = url_1.parse(fileUrl);
        var requesters = {
            'http:': http_1.request,
            'https:': https_1.request
        };
        return new Promise(function (resolve, reject) {
            var requester = requesters[urlParts.protocol];
            if (requester) {
                var req = requester({
                    hostname: urlParts.hostname,
                    port: urlParts.port,
                    path: urlParts.path,
                    method: 'GET'
                }, function (res) {
                    var responseData = '';
                    res.setEncoding('utf8');
                    res.on('data', function (chunk) {
                        responseData += chunk.toString();
                    });
                    res.on('end', function () {
                        resolve(_this.getObjectFromString(responseData));
                    });
                });
                req.on('error', function (e) {
                    reject("problem with request: " + e.message);
                });
                req.end();
            }
            else {
                reject('unable to identify url protocol');
            }
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getObjectFromString = function (data) {
        return new Promise(function (resolve, reject) {
            try {
                resolve(JSON.parse(data.toString()));
            }
            catch (e) {
                reject(e);
            }
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getObjectFromFile = function (filePath) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            fs_1.readFile(filePath, function (err, data) {
                if (err) {
                    reject(err);
                }
                else {
                    resolve(_this.getObjectFromString(data));
                }
            });
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.resume = function () {
        return this.domains.Debugger.resume();
    };
    ChromeDebuggingProtocolDebugger.prototype.pause = function () {
        return this.domains.Debugger.pause();
    };
    ChromeDebuggingProtocolDebugger.prototype.stepOver = function () {
        return this.domains.Debugger.stepOver();
    };
    ChromeDebuggingProtocolDebugger.prototype.stepInto = function () {
        return this.domains.Debugger.stepInto();
    };
    ChromeDebuggingProtocolDebugger.prototype.stepOut = function () {
        return this.domains.Debugger.stepOut();
    };
    ChromeDebuggingProtocolDebugger.prototype.getProperties = function (params) {
        return this.domains.Runtime.getProperties(params);
    };
    ChromeDebuggingProtocolDebugger.prototype.evaluateOnFrames = function (expression, frames) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (frames.length > 0) {
                var frame = frames.shift();
                if (frame && frame.callFrameId) {
                    _this
                        .domains
                        .Debugger
                        .evaluateOnCallFrame({
                        callFrameId: frame.callFrameId,
                        expression: expression,
                        generatePreview: false,
                        silent: true,
                        returnByValue: false,
                        includeCommandLineAPI: false
                    })
                        .then(function (result) {
                        var lookOnParent = frames.length > 0 &&
                            result.result.subtype === 'error' &&
                            result.result.className !== 'SyntaxError';
                        if (lookOnParent) {
                            resolve(_this.evaluateOnFrames(expression, frames));
                        }
                        else if (result && !result.exceptionDetails) {
                            resolve(result);
                        }
                        else {
                            reject(result);
                        }
                    });
                }
                else {
                    reject('frame has no id');
                }
            }
            else {
                reject('there are no frames to evaluate');
            }
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.evaluate = function (expression) {
        var frames = (this.callFrames || []).slice();
        return this.evaluateOnFrames(expression, frames);
    };
    ChromeDebuggingProtocolDebugger.prototype.getScriptById = function (scriptId) {
        return this.scripts.find(function (s) {
            return parseInt(s.scriptId) === scriptId;
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getScriptByUrl = function (url) {
        return this.scripts.find(function (s) {
            return s.url === url;
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getCallStack = function () {
        var _this = this;
        return this.callFrames
            .filter(function (frame) {
            frame.location.script = _this.getScriptById(parseInt(frame.location.scriptId));
            var sourceMap = lodash_1.get(frame, 'location.script.sourceMap');
            if (sourceMap) {
                var position = sourceMap.getOriginalPosition(frame.location.lineNumber, parseInt(frame.location.columnNumber));
                if (position) {
                    frame.location.script.url = position.url;
                    frame.location.lineNumber = position.lineNumber;
                    frame.location.columnNumber = position.columnNumber;
                    return true;
                }
                else {
                    return false;
                }
            }
            return true;
        })
            .map(function (frame) {
            return {
                name: frame.functionName,
                columnNumber: frame.location.columnNumber,
                lineNumber: frame.location.lineNumber,
                filePath: frame.location.script.url,
                scope: _this.getScopeFromFrame(frame)
            };
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getFrameByIndex = function (index) {
        return this.callFrames[index];
    };
    ChromeDebuggingProtocolDebugger.prototype.setBreakpointFromScript = function (script, lineNumber, condition) {
        var _this = this;
        return new Promise(function (resolve) {
            var position = {
                url: script.url,
                lineNumber: lineNumber,
                condition: ''
            };
            if (script.sourceMap && script.sourceMap.getPosition) {
                position = script.sourceMap.getPosition(lineNumber);
            }
            position.url = _this.getFilePathFromUrl(position.url);
            position.condition = condition;
            _this
                .domains
                .Debugger
                .setBreakpointByUrl(position)
                .then(function (breakpoint) {
                _this.breakpoints.push({
                    id: breakpoint.breakpointId,
                    url: script.url,
                    columnNumber: 0,
                    condition: condition,
                    lineNumber: lineNumber
                });
                resolve(breakpoint);
            })
                .catch(function (message) {
                // do nothing
            });
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.addBreakpoint = function (url, lineNumber, condition) {
        var _this = this;
        return this
            .removeBreakpoint(url, lineNumber)
            .then(function () {
            return new Promise(function (resolve, reject) {
                var script = _this.getScriptByUrl(url);
                if (script) {
                    resolve(_this.setBreakpointFromScript(script, lineNumber, condition));
                }
                else {
                    reject(url + " is not parsed");
                }
            });
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getBreakpointById = function (id) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var found = _this.breakpoints.find(function (b) {
                return (b.id === id);
            });
            resolve(found);
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getBreakpoint = function (url, lineNumber) {
        return this.breakpoints.find(function (b) {
            return (b.url === url && b.lineNumber === lineNumber);
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.removeBreakpoint = function (url, lineNumber) {
        var breakpoint = this.getBreakpoint(url, lineNumber);
        if (breakpoint) {
            var index = this.breakpoints.indexOf(breakpoint);
            this.breakpoints.splice(index, 1);
            return this.domains.Debugger.removeBreakpoint({
                breakpointId: breakpoint.id
            });
        }
        return Promise.resolve();
    };
    ChromeDebuggingProtocolDebugger.prototype.getScopeFromFrame = function (frame) {
        var scope = frame.scopeChain.slice();
        if (frame.this) {
            scope.unshift({
                type: 'this',
                object: frame.this
            });
        }
        return scope.map(function (s) {
            return {
                name: s.type,
                value: s.object
            };
        });
    };
    ChromeDebuggingProtocolDebugger.prototype.getScope = function () {
        var firstFrame = this.getFrameByIndex(0);
        return this.getScopeFromFrame(firstFrame);
    };
    // Events
    ChromeDebuggingProtocolDebugger.prototype.didClose = function (cb) {
        this.events.addListener('didClose', cb);
    };
    ChromeDebuggingProtocolDebugger.prototype.didLogMessage = function (cb) {
        this.events.addListener('didLogMessage', cb);
    };
    ChromeDebuggingProtocolDebugger.prototype.didThrownException = function (cb) {
        this.events.addListener('didThrownException', cb);
    };
    ChromeDebuggingProtocolDebugger.prototype.didLoadScript = function (cb) {
        this.events.addListener('didLoadScript', cb);
    };
    ChromeDebuggingProtocolDebugger.prototype.didPause = function (cb) {
        this.events.addListener('didPause', cb);
    };
    ChromeDebuggingProtocolDebugger.prototype.didResume = function (cb) {
        this.events.addListener('didResume', cb);
    };
    return ChromeDebuggingProtocolDebugger;
}());
exports.ChromeDebuggingProtocolDebugger = ChromeDebuggingProtocolDebugger;
